okhttp/src/main/java/io/grpc/okhttp/OkHttpProtocolNegotiator.java

```diff
@@ -152,31 +179,61 @@ public String negotiate(SSLSocket sslSocket, String hostname, List<Protocol> pro
     /**
      * Override {@link Platform}'s configureTlsExtensions for Android older than 5.0, since OkHttp
      * (2.3+) only support such function for Android 5.0+.
+     *
+     * <p>Note: Prior to Android Q, the standard way of accessing some Conscrypt features was to
+     * use reflection to call hidden APIs. Beginning in Q, there is public API for all of these
+     * features. We attempt to use the public API where possible, but also still call the
+     * hidden versions to continue to support old versions of Conscrypt that might be bundled with
+     * apps or third-party TLS providers that might have taken advantage of being able to
+     * duck-type their way into compatibility.
      */
     @Override
     protected void configureTlsExtensions(
         SSLSocket sslSocket, String hostname, List<Protocol> protocols) {
       // Enable SNI and session tickets.
       if (hostname != null) {
-        SET_USE_SESSION_TICKETS.invokeOptionalWithoutCheckedException(sslSocket, true);
+        try {
+          if (sslSocketsIsSupportedSocket != null
+              && (boolean) sslSocketsIsSupportedSocket.invoke(null, sslSocket)) {
+            sslSocketsSetUseSessionTickets.invoke(null, sslSocket, true);
+          } else {
+            SET_USE_SESSION_TICKETS.invokeOptionalWithoutCheckedException(sslSocket, true);
+          }
+        } catch (IllegalAccessException ignored) {
+        } catch (InvocationTargetException ignored) {
+        }
         SET_HOSTNAME.invokeOptionalWithoutCheckedException(sslSocket, hostname);
       }
 
-      Object[] parameters = {Platform.concatLengthPrefixed(protocols)};
-      if (platform.getTlsExtensionType() == TlsExtensionType.ALPN_AND_NPN) {
-        SET_ALPN_PROTOCOLS.invokeWithoutCheckedException(sslSocket, parameters);
-      }
-
-      if (platform.getTlsExtensionType() != TlsExtensionType.NONE) {
-        SET_NPN_PROTOCOLS.invokeWithoutCheckedException(sslSocket, parameters);
-      } else {
-        throw new RuntimeException("We can not do TLS handshake on this Android version, please"
-            + " install the Google Play Services Dynamic Security Provider to use TLS");
+      // Enable ALPN and NPN if necessary.
+      SSLParameters sslParams = sslSocket.getSSLParameters();
+      SET_APPLICATION_PROTOCOLS
+          .invokeOptionalWithoutCheckedException(sslParams, (Object) protocolIds(protocols));
+      sslSocket.setSSLParameters(sslParams);
+      if (!isPlatformSocket(sslSocket) || !SET_APPLICATION_PROTOCOLS.isSupported(sslParams)) {
```

Comment:
It is strange that the code invokes SET_APPLICATION_PROTOCOLS _and then_ checks to see if it is supported. Is this because we don't know which API will work?

I see you said:
> We need to use these new APIs whenever possible while still calls old versions whenever it does not break things.

But it didn't say _why_.

Reply:
The logic of calling which API is subtle here (even myself need to think for a couple of seconds why it branches in this way).

The bottomline is, we must not invoke SET_ALPN_PROTOCOLS/SET_NPN_PROTOCOLS when _it is Android R_ and _it is using Android's Conscrypt (i.e., `com.android.org.conscrypt`)_. Otherwise, it breaks in internal version of Android R.

Other than that, we do both. This preserves compatibility for the most extent while avoid the disallowed invocation in Android R's Conscrypt. 

`!isPlatformSocket(sslSocket)` checks if it uses Android OS Conscrypt and `!SET_APPLICATION_PROTOCOLS.isSupported(sslParams)` checks if it is in Android Q+. Avoid calling SET_ALPN_PROTOCOLS/SET_NPN_PROTOCOLS when both conditions are met. Otherwise, it is ok to do it.